### ISSUE TITLE
(Client=55) Flush memory during reindex so it happens faster

### ISSUE THREAD

**Issue by garrett-gottlieb** [orientechnologies/orientdb#3308]
This is coming specifically from a scaled database rebuild with 100M+ records. The reindex on an incredibly powerful server (120GB RAM 800GB SSD) is only going at 3-10000 items/sec. Initially, it was going faster (35000 items/sec) but it quickly slowed down once the memory filled up. This is because all previous indexes are still in memory, so the newest reindex has very little to play with. If memory was flushed during each reindex, the reindex would happen much faster and stop clogging up database operations.

Here is the memory during indexing:

```
             total       used       free     shared    buffers     cached
Mem:          120G       119G       648M       392K       148M        55G
-/+ buffers/cache:        63G        56G
Swap:           0B         0B         0B
```

It stays like this, even when a new index starts. It should free most of the memory when an index starts.

The reindex on our database seems like it could take 5+ hours, and due to the frequency of these (and potential occurrence during production), minimizing downtime is extremely crucial.

**Comment by laa** [comment]
Hi,
I do not think that reindex itself should be used in production at all.
Because it leads to errors in query results.

I think we should resolve root of this problem instead of reindex problem.
So my question is why do you have reindex, is it because of server crash ?

**Comment by garrett-gottlieb** [comment]
Yes, I've logged separate issues whenever the server crashes. The lase one was due to an unindexed query that caused a lock timeout.

This is not just a reindex. Building a new index can is also slow. Speeding this up speeds up development immensely, and recovering from crashes which is an unfortunate inevitability.

**Comment by laa** [comment]
We need to understand why do you have index rebuild if you use graph database. Do you use non tx modification operations during filling of database content ?

**Comment by laa** [comment]
To flush or not to flush cache is highly specific thing, for example if you do index rebuild and do a lot of other operations in parallel other db users will not be happy that the rest of operations are suddenly became noticeably slower. So I will create ```purgedcache``` command which you may call if you need it. But lets find out and fix root problem.

**Comment by garrett-gottlieb** [comment]
We do use non-tx operations for simple updates and inserts. Is it necessary to switch all of those to tx?

Generally speaking, we will take down a node in the cluster, build the index, and then put it back in.

For this purpose, could you call the `purgeCache` command after an index is rebuilt during a database rebuild? I hope that would speed it up.

**Comment by lvca** [comment]
@garrett-gottlieb If you always use transactions you're protected by such problems.

**Comment by laa** [comment]
Hi @garrett-gottlieb . Your case is not general case. I may provide command so you will call it before index rebuild. But I do not think that it should be always called.

**Comment by laa** [comment]
Fixed in  3.1 we group all adjacent pages together. Also, we use async io to trigger up to 1K write requests simultaneously.
